### PR TITLE
docker: fix docker file image and apply some good practices

### PR DESCRIPTION
**Pull request by cpanato**
#### Summary
Someone setup in docker hub the automated builds when we have a new merge to the main branch, this build is broken for quite some time now.

This PR fix this and refactor a bit the dockerfile to add some good practices.

- Instead of cloning the repo we use the repo it self, otherwise in the automated build when building for a TAG it will use the main branch always and not the TAG
- uses distroless image to reduce the scope of things installed
- using tagged images to better know which versions we are using to build it

with those fixes I've push a test image to this place: ` ctadeu/focalboard:test` that you can pull and test it

```
$ docker pull  ctadeu/focalboard:test
$ docker run -it -v "/tmp/focalboard-data:/data" -p 80:8000 ctadeu/focalboard:test
```

Also notice the docker compose configs are not working, will fix that in a follow up, was not the focus here

### CODE REVIEW

**Comment by cpanato** [comment]
for referencing the docker automated builds

Screenshot 2021-06-09 at 15 07 03

File: Dockerfile.build [hunk 1]
```diff
@@ -7,7 +7,7 @@
 # docker cp test:/dist/focalboard-server-linux-amd64.tar.gz .
 
 # build frontend
-FROM node:16.1.0 AS frontend
+FROM node:16.3.0 AS frontend
```
Comment (harshilsharma63) [inline Dockerfile.build:10]: We should always keep this in sync with `webapp/.nvmrc`. Can you update the same in there as well?

File: docker/Dockerfile [hunk 2]
```diff
@@ -1,40 +1,34 @@
...
-    apk add ca-certificates gcc git make musl-dev && \
-    git clone https://github.com/mattermost/focalboard && \
-    cd focalboard && \
-    make server-linux
+WORKDIR /webapp
+ADD webapp/ /webapp
 
-FROM node:alpine as nodebuild
+RUN npm install --no-optional && \
+    npm run pack
```
Comment (harshilsharma63) [inline docker/Dockerfile:8]: I think we should avoid directly running npm commands and instead use `Makefile` as the entry point and interface for all things DevOps.

For example -

```
RUN npm install --no-optional &&\
cd .. && make webapp
```

This allows us to encapsulate the compilation and build process and avoid frequent change to DevOps scripts. This should be discussed in the Focalboard channel as well.

**Review by harshilsharma63 — CHANGES_REQUESTED** [review]

File: .dockerignore [hunk 3]
```diff
@@ -0,0 +1,8 @@
+CHANGELOG.md
+README.md
+
+.github/
+mac/
+win-wpf/
+mattermost-plugin/
+website/
```
Comment (jespino) [inline .dockerignore:8]: probably we can also ad here `linux/` directory that is for generating the "linux desktop app"

File: docker/Dockerfile (hunk 2 above)
Comment (jespino) [reply docker/Dockerfile:8]: I think both options are ok, actually I almost prefer to not have the Makefile make the frontend and the backend work independently, but I understand that is really handy for others. I think the webapp should be compilable without the makefile with one right command.

**Review by jespino — APPROVED** [review]
LGTM. An improvement suggestion.

**Comment by jmcph4** [comment]
So, currently on [Debian Buster](https://github.com/mattermost/focalboard/issues/574), this branch builds fine but then falls over? Just trying to determine if this is due to misconfiguration on my part or if there are further issues with the Docker releases.

```sh
$ git clone git@github.com:mattermost/focalboard.git
$ cd focalboard
$ git fetch origin fix-docker
$ git checkout fix-docker
$ sudo docker build -t kanban -f docker/Dockerfile .
Sending build context to Docker daemon  14.01MB
Step 1/18 : FROM node:16.3.0 as nodebuild
 ---> d1b3088a17b1
Step 2/18 : WORKDIR /webapp
 ---> Using cache
 ---> f9117f63adc1
Step 3/18 : ADD webapp/ /webapp
 ---> 88c104e00faf
Step 4/18 : RUN npm install --no-optional &&     npm run pack
 ---> Running in 8619ab152fac

added 1417 packages, and audited 1418 packages in 2m

22 packages are looking for funding
  run `npm fund` for details

26 vulnerabilities (3 moderate, 23 high)

To address issues that do not require attention, run:
  npm audit fix

To address all issues (including breaking changes), run:
  npm audit fix --force

Run `npm audit` for details.
npm notice 
npm notice New minor version of npm available! 7.15.1 -> 7.17.0
npm notice Changelog: <https://github.com/npm/cli/releases/tag/v7.17.0>
npm notice Run `npm install -g npm@7.17.0` to update!
npm notice 

> focalboard@0.6.7 pack
> NODE_ENV=production webpack --config webpack.js

assets by chunk 204 KiB (auxiliary name: main)
  asset static/compass-icons.svg 76.8 KiB [emitted] [from: src/fonts/compass-icons/compass-icons.svg] (auxiliary name: main)
  asset static/compass-icons.eot 42.6 KiB [emitted] [from: src/fonts/compass-icons/compass-icons.eot] (auxiliary name: main)
  asset static/compass-icons.ttf 42.4 KiB [emitted] [from: src/fonts/compass-icons/compass-icons.ttf] (auxiliary name: main)
  asset static/compass-icons.woff 23.4 KiB [emitted] [from: src/fonts/compass-icons/compass-icons.woff] (auxiliary name: main)
  asset static/compass-icons.woff2 19.2 KiB [emitted] [from: src/fonts/compass-icons/compass-icons.woff2] (auxiliary name: main)
asset static/main.js 2.19 MiB [emitted] [minimized] [big] (name: main)
asset static/easymde.min.css 12.1 KiB [emitted] [from: node_modules/easymde/dist/easymde.min.css] [copied]
asset static/favicon.svg 2.36 KiB [emitted] [from: static/favicon.svg] [copied]
asset index.html 503 bytes [emitted]
orphan modules 1.32 MiB [orphan] 258 modules
runtime modules 3.56 KiB 9 modules
javascript modules 4.25 MiB
  modules by path ./src/ 657 KiB 350 modules
  modules by path ./node_modules/ 3.61 MiB 212 modules
  fs (ignored) 15 bytes [built] [code generated]
json modules 74.4 KiB 11 modules

WARNING in asset size limit: The following asset(s) exceed the recommended size limit (244 KiB).
This can impact web performance.
Assets: 
  static/main.js (2.19 MiB)

WARNING in entrypoint size limit: The following entrypoint(s) combined asset size exceeds the recommended limit (244 KiB). This can impact web performance.
Entrypoints:
  main (2.19 MiB)
      static/main.js

WARNING in webpack performance recommendations: 
You can limit the size of your bundles by using import() or require.ensure to lazy load some parts of your application.
For more info visit https://webpack.js.org/guides/code-splitting/

webpack 5.28.0 compiled with 3 warnings in 88484 ms
npm notice 
npm notice New minor version of npm available! 7.15.1 -> 7.17.0
npm notice Changelog: <https://github.com/npm/cli/releases/tag/v7.17.0>
npm notice Run `npm install -g npm@7.17.0` to update!
npm notice 
Removing intermediate container 8619ab152fac
 ---> a90ea88a103a
Step 5/18 : FROM golang:1.16.5 as gobuild
1.16.5: Pulling from library/golang
d960726af2be: Already exists 
e8d62473a22d: Already exists 
8962bc0fad55: Already exists 
65d943ee54c1: Already exists 
f2253e6fbefa: Pull complete 
186c77a2a533: Pull complete 
db807893dccf: Pull complete 
Digest: sha256:74681bd28112ca96a4070c8e5c44747de14042108c22774d8f38bf488fb02838
Status: Downloaded newer image for golang:1.16.5
 ---> b09f7387a719
Step 6/18 : WORKDIR /go/src/focalboard
 ---> Running in 9ff10680b154
Removing intermediate container 9ff10680b154
 ---> d5255a2170d8
Step 7/18 : ADD . /go/src/focalboard
 ---> 246d82f8d3a1
Step 8/18 : RUN  make server-linux
 ---> Running in 0c05cb2e1042
mkdir -p bin/linux
cd server; env GOOS=linux GOARCH=amd64 go build -ldflags '-X "github.com/mattermost/focalboard/server/model.BuildNumber=dev" -X "github.com/mattermost/focalboard/server/model.BuildDate=Mon Jun 14 06:34:24 UTC 2021" -X "github.com/mattermost/focalboard/server/model.BuildHash=18b2738d6bd2ab1b810e0e61ff75d07601eb4f39" -X "github.com/mattermost/focalboard/server/model.Edition=linux"' -o ../bin/linux/focalboard-server ./main
go: downloading github.com/google/uuid v1.2.0
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/mattermost/mattermost-server/v5 v5.3.2-0.20210524045451-a4f7df6f6e3c
go: downloading github.com/oklog/run v1.1.0
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/spf13/viper v1.7.1
go: downloading github.com/mattermost/logr v1.0.13
go: downloading github.com/mattermost/logr/v2 v2.0.2
go: downloading github.com/prometheus/client_golang v1.10.0
go: downloading github.com/Masterminds/squirrel v1.5.0
go: downloading github.com/go-sql-driver/mysql v1.5.0
go: downloading github.com/golang-migrate/migrate/v4 v4.14.1
go: downloading github.com/lib/pq v1.10.0
go: downloading github.com/mattn/go-sqlite3 v2.0.3+incompatible
go: downloading github.com/rudderlabs/analytics-go v3.3.1+incompatible
go: downloading github.com/gorilla/websocket v1.4.2
go: downloading github.com/minio/minio-go/v7 v7.0.10
go: downloading github.com/stretchr/testify v1.7.0
go: downloading github.com/fsnotify/fsnotify v1.4.9
go: downloading github.com/hashicorp/hcl v1.0.0
go: downloading github.com/magiconair/properties v1.8.5
go: downloading github.com/mitchellh/mapstructure v1.4.1
go: downloading github.com/pelletier/go-toml v1.8.1
go: downloading github.com/spf13/afero v1.6.0
go: downloading github.com/spf13/cast v1.3.1
go: downloading github.com/spf13/jwalterweatherman v1.1.0
go: downloading github.com/spf13/pflag v1.0.5
go: downloading github.com/subosito/gotenv v1.2.0
go: downloading gopkg.in/ini.v1 v1.62.0
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading github.com/wiggin77/merror v1.0.3
go: downloading github.com/francoispqt/gojay v1.2.13
go: downloading golang.org/x/crypto v0.0.0-20210322153248-0c34fe9e7dc2
go: downloading github.com/beorn7/perks v1.0.1
go: downloading github.com/cespare/xxhash v1.1.0
go: downloading github.com/cespare/xxhash/v2 v2.1.1
go: downloading github.com/golang/protobuf v1.5.1
go: downloading github.com/prometheus/client_model v0.2.0
go: downloading github.com/prometheus/common v0.20.0
go: downloading github.com/prometheus/procfs v0.6.0
go: downloading github.com/lann/builder v0.0.0-20180802200727-47ae307949d0
go: downloading github.com/hashicorp/go-multierror v1.1.1
go: downloading github.com/segmentio/backo-go v0.0.0-20200129164019-23eae7c10bd3
go: downloading github.com/tidwall/gjson v1.7.3
go: downloading github.com/xtgo/uuid v0.0.0-20140804021211-a0b114877d4c
go: downloading github.com/wiggin77/srslog v1.0.1
go: downloading go.uber.org/zap v1.16.0
go: downloading gopkg.in/natefinch/lumberjack.v2 v2.0.0
go: downloading github.com/json-iterator/go v1.1.10
go: downloading github.com/minio/md5-simd v1.1.2
go: downloading github.com/minio/sha256-simd v1.0.0
go: downloading golang.org/x/net v0.0.0-20210316092652-d523dce5a7f4
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/blang/semver v3.5.1+incompatible
go: downloading github.com/disintegration/imaging v1.6.2
go: downloading github.com/dyatlov/go-opengraph v0.0.0-20210112100619-dae8665a5b09
go: downloading github.com/mattermost/ldap v0.0.0-20201202150706-ee0e6284187d
go: downloading github.com/pborman/uuid v1.2.1
go: downloading github.com/tinylib/msgp v1.1.5
go: downloading golang.org/x/text v0.3.5
go: downloading github.com/mattermost/go-i18n v1.11.0
go: downloading golang.org/x/sys v0.0.0-20210324051608-47abb6519492
go: downloading google.golang.org/protobuf v1.26.0
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.1
go: downloading github.com/lann/ps v0.0.0-20150810152359-62de8c46ede0
go: downloading github.com/hashicorp/errwrap v1.1.0
go: downloading github.com/tidwall/match v1.0.3
go: downloading github.com/tidwall/pretty v1.1.0
go: downloading github.com/wiggin77/cfg v1.0.2
go: downloading go.uber.org/atomic v1.7.0
go: downloading go.uber.org/multierr v1.6.0
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/modern-go/reflect2 v1.0.1
go: downloading github.com/klauspost/cpuid v1.3.1
go: downloading github.com/klauspost/cpuid/v2 v2.0.5
go: downloading github.com/rs/xid v1.2.1
go: downloading golang.org/x/image v0.0.0-20210220032944-ac19c3e999fb
go: downloading github.com/go-asn1-ber/asn1-ber v1.5.3
go: downloading github.com/philhofer/fwd v1.1.1
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/pmezard/go-difflib v1.0.0
go: downloading gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b
Removing intermediate container 0c05cb2e1042
 ---> e43a7807bda2
Step 9/18 : FROM gcr.io/distroless/base-debian10
latest: Pulling from distroless/base-debian10
5dea5ec2316d: Pull complete 
bb771d6dc9a1: Pull complete 
Digest: sha256:6ec6da1888b18dd971802c2a58a76a7702902b4c9c1be28f38e75e871cedc2df
Status: Downloaded newer image for gcr.io/distroless/base-debian10:latest
 ---> a4cf6da932ac
Step 10/18 : WORKDIR /opt/focalboard
 ---> Running in a9f929a35a74
Removing intermediate container a9f929a35a74
 ---> 0f922ecc9f07
Step 11/18 : COPY --from=nodebuild /webapp/pack pack/
 ---> d823d8b812af
Step 12/18 : COPY --from=gobuild /go/src/focalboard/bin/linux/focalboard-server bin/
 ---> 395a7af90ad1
Step 13/18 : COPY --from=gobuild /go/src/focalboard/LICENSE.txt LICENSE.txt
 ---> 16a0caf5248b
Step 14/18 : COPY --from=gobuild /go/src/focalboard/docker/server_config.json config.json
 ---> 1f2e411e8f12
Step 15/18 : USER nobody
 ---> Running in 7f86fb4392e0
Removing intermediate container 7f86fb4392e0
 ---> 04370dcd3361
Step 16/18 : EXPOSE 8000/tcp
 ---> Running in 35d83f6018d0
Removing intermediate container 35d83f6018d0
 ---> 8194aa452236
Step 17/18 : VOLUME /data
 ---> Running in 7381121cffc5
Removing intermediate container 7381121cffc5
 ---> 864cd3c819f1
Step 18/18 : CMD ["/opt/focalboard/bin/focalboard-server"]
 ---> Running in e34e3a4e1462
Removing intermediate container e34e3a4e1462
 ---> 551b50f8911b
Successfully built 551b50f8911b
Successfully tagged kanban:latest
$ sudo docker run -it -v "/home/[REDACTED]/focalboard-data:/data" -p 80:8000 kanban
2021/06/14 06:38:43 readConfigFile
2021/06/14 06:38:43 {ServerRoot:http://localhost:8000 Port:8000 DBType:sqlite3 DBConfigString:/data/focalboard.db DBTablePrefix: UseSSL:false SecureCookie:false WebPath:./pack FilesDriver:local FilesS3Config:{AccessKeyId: SecretAccessKey: Bucket: PathPrefix: Region: Endpoint: SSL:false SignV2:false SSE:false Trace:false} FilesPath:/data/files Telemetry:true PrometheusAddress: WebhookUpdate:[] Secret: SessionExpireTime:2592000 SessionRefreshTime:18000 LocalOnly:false EnableLocalMode:true LocalModeSocketLocation:/var/tmp/focalboard_local.socket AuthMode:native LoggingFile: LoggingEscapedJson:}
$ echo $?
1
```

**Review by chenilim — APPROVED** [review]
Thanks @cpanato! I'm going to have to defer to others to review the build config. LGTM.
This might be a separate issue, but the dockerfile should ultimately install the compiled release binaries, which are licensed differently among other things.

File: Dockerfile.build (hunk 1 above)
Comment (cpanato) [reply Dockerfile.build:10]: done

File: docker/Dockerfile (hunk 2 above)
Comment (cpanato) [reply docker/Dockerfile:8]: i will keep as-is for now

**Comment by cpanato** [comment]
@jmcph4 works for me i just tried

did you create the directory `/home/[REDACTED]/focalboard-data` ? 

> So, currently on [Debian Buster](https://github.com/mattermost/focalboard/issues/574), this branch builds fine but then falls over? Just trying to determine if this is due to misconfiguration on my part or if there are further issues with the Docker releases.

**Review by harshilsharma63 — APPROVED** [review]

File: webapp/.nvmrc [hunk 4]
```diff
@@ -1 +1 @@
-v16.1.0
+v16.3.0
```
Comment (kamre) [inline webapp/.nvmrc:1]: @cpanato should workflow files for github actions [like this one](https://github.com/mattermost/focalboard/blob/main/.github/workflows/ci.yml#L32) that install node be also updated to use the new version?

Comment (cpanato) [reply webapp/.nvmrc:1]: sgtm

**Pull request merged**